jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/stats/TimeBasedSlidingWindow.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright @ 2018 - present 8x8, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.utils.stats
+
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+
+/**
+ * Maintain a queue of values which fall inside the given window size.
+ *
+ * Eviction is done after adding an element in [add], and can be forced
+ * via [forceEviction].  Any time an element is evicted, the [evictionHandler]
+ * method is called with the evicted value.  Forcing eviction before any access
+ * is done is required, as it's possible there are stale values in the collection.
+ */
+open class TimeBasedSlidingWindow<T : Any>(
+    private val windowSize: Duration,
+    private val evictionHandler: (T) -> Unit = {},
+    private val clock: Clock = Clock.systemUTC()
+) {
+    private val queue = java.util.ArrayDeque<TimeEntry>()
+
+    private val evictionPredicate: (TimeEntry) -> Boolean = { entry ->
```

Comment:
Now that you've refactored this, should evictionPredicate still be a `val` rather than just a `fun`?

Reply:
Defining a member function for this felt too "heavy", and the predicate itself felt a bit too complicated too inline it in the eviction method, so that's how I ended up here.

Reply:
I'd think a `val` predicate would be heavier-weight than a member function, if you look at how it's actually instantiated.

Reply:
It also just feels odd to me to have this as a member function. This is basically a `Predicate` instance, which we'd declare as a member field, not function.